--- origsrc/pygobject-2.18.0/configure.ac	2009-08-23 14:46:30.013700100 -0500
+++ src/pygobject-2.18.0/configure.ac	2009-08-23 14:46:18.497274100 -0500
@@ -164,14 +164,15 @@ AC_ARG_WITH(ffi,
 if test x"$with_libffi" = xno ; then
   have_libffi=false
 else
-  PKG_CHECK_MODULES(FFI, libffi >= 3.0, have_libffi=yes, have_libffi=no)
+  PKG_CHECK_MODULES(FFI, libffi >= 3.0, [LIBFFI_PC=libffi; have_libffi=yes],[
+    AC_CHECK_LIB(ffi, ffi_call, [FFI_LIBS=-lffi; have_libffi=yes], have_libffi=no)],
+  ])
 fi
 if test x"$with_libffi" = xyes && test x"$have_libffi" = xno ; then
   AC_MSG_ERROR([ffi requested, but not found])
 fi
 if test x"$have_libffi" = xyes; then
   AC_DEFINE(HAVE_FFI_H,1,[Have libffi include files])
-  LIBFFI_PC=libffi
 fi
 AC_MSG_RESULT([$have_libffi])
 AM_CONDITIONAL(HAVE_LIBFFI, test "$have_libffi" = "yes")
